yjit/src/codegen.rs

```diff
@@ -4677,15 +4677,18 @@ fn gen_branchnil(
         asm.cmp(val_opnd, Opnd::UImm(Qnil.into()));
         // Generate the branch instructions
         let ctx = asm.ctx;
-        gen_branch(
+        if gen_branch(
             jit,
             asm,
             jump_block,
             &ctx,
             Some(next_block),
             Some(&ctx),
             BranchGenFn::BranchNil(Cell::new(BranchShape::Default)),
-        )?;
+        ).is_none() {
+            // Block unusable without the branch.
+            jit.block_abandoned = true;
```

Comment:
It seems like we always set this flag when `gen_branch()` returns `None`. Similarly to `jit.defer_compilation()`, could we introduce `jit.gen_branch()` that sets `jit.block_abandoned = true;` on failure (or let `gen_branch()` do that)?